designs/2024-baseline-support/README.md

```diff
@@ -0,0 +1,379 @@
+
+Note that the error detection in `cli.js` occurs before the error counting. This allow us to update the suppressions file and modify the errors, before it is time to count errors. Please refer to the last example of the "Implementation notes" for more details.
+
+Furthermore, ESLint cache (`--cache`) must include the full list of detected violations, even those in the suppressions file. This approach has the following benefits:
+
+- Generating the suppressions file can be based on the cache file and should be faster when the cache file is used.
+- Allows developers to update the suppressions file and then re-lint still taking the same cache into consideration.
+- It even allows developers to delete the suppressions file and still take advantage of the cached file in subsequent runs. 
+
+### Implementation notes
+
+To introduce the above-mentioned options, we will need to: 
+
+* add the new options in `default-cli-options.js`.
+* adjust the config for optionator.
+* add the new options as comments and arguments for eslint.
+* update documentation to explain the newly introduced feature.
+
+A new type must be created to represent the suppressions file:
+
+``` js
+/**
+ * @typedef {Record<string, Record<string, { count: number }>} SuppressedViolations
+ */
+```
+
+A new class must be created to manage the suppressions file:
+
+``` js
+class SuppressedViolationsManager {
+    /**
+     * Creates a new instance of SuppressedViolationsManager.
+     * @param {string} suppressionsLocation The location of the suppressions file.
+     */
+    constructor(suppressionsLocation) {}
+
+    /**
+     * Updates the suppressions file based on the current violations.
+     * 
+     * @param {LintResult[]} results The lint results.
+     * @returns {void}
+     */
+    suppressAll(results) 
+
+    /**
+     * Updates the suppressions file based on the current violations and the provided rule.
+     * 
+     * @param {LintResult[]} results The lint results.
+     * @param {string[]} rules The rules to suppress.
+     * @returns {void}
+     */
+    suppressRules(results, rules)
+
+    /**
+     * Removes old suppressions that do not occur anymore.
+     * @returns {void}
+     */
+    prune()
+
+    /**
+     * Checks the provided suppressions against the lint results.
+     * 
+     * For each error included in `results`, checks if the error and the file are in `suppressions`.
+     *  If yes, the count is decreased by 1 and ignores the error unless the count has reached zero. 
+     * Otherwise, it keeps the error.
+     * 
+     * It returns the lint results that are not in the suppressions file,
+     * as well as the unmatched suppressions.
+     * 
+     * @param {LintResult[]} results The lint results.
+     * @param {SuppressedViolations} suppressions The suppressions.
+     * @returns {{
+     *   results: LintResult[],
+     *   unmatched: SuppressedViolations
+     * }}
+     */
+    applySuppressions(results, suppressions)
```

Comment:
Thanks for the update. Maybe I'm misunderstanding something, so bear with me. It would be helpful if you could add a draft implementation to show how this function will work.

To clarify, this is an example I would like to understand. Suppose you have a file test.js where the `no-undef` rule would normally report three errors:

```
/* eslint no-undef: "error" */

a = b + c;
```

[**Playground link**](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVuZGVmOiBcImVycm9yXCIgKi9cblxuYSA9IGIgKyBjOyIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsic291cmNlVHlwZSI6Im1vZHVsZSIsInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX19fX0=)

What will the function return when the suppression file specifies a `count` of 2? This is what the arguments will look like:

```js
const results = [{
    filePath: '/my-project/test.js',
    messages: [{
        ruleId: 'no-undef',
        severity: 2,
        message: "'a' is not defined.",
        line: 3,
        column: 1,
        nodeType: 'Identifier',
        messageId: 'undef',
        endLine: 3,
        endColumn: 2
    },
    {
        ruleId: 'no-undef',
        severity: 2,
        message: "'b' is not defined.",
        line: 3,
        column: 5,
        nodeType: 'Identifier',
        messageId: 'undef',
        endLine: 3,
        endColumn: 6
    },
    {
        ruleId: 'no-undef',
        severity: 2,
        message: "'c' is not defined.",
        line: 3,
        column: 9,
        nodeType: 'Identifier',
        messageId: 'undef',
        endLine: 3,
        endColumn: 10
    }],
    suppressedMessages: [],
    errorCount: 3,
    fatalErrorCount: 0,
    warningCount: 0,
    fixableErrorCount: 0,
    fixableWarningCount: 0,
    source: '/* eslint no-undef: "error" */\n\na = b + c;\n',
    usedDeprecatedRules: []
}];

const suppressions = {
    "test.js": {
        "no-undef": {
            count: 2
        }
    }
};
```

You can see that the input `results` contains details about each error, while the suppressions are just numbers. What will the call to `applySuppressions(results, suppressions)` return in this case?

Reply:
In that case it will return an updated copy of results and an empty object for unmatched (because all suppressions are matched). 

``` js
// results
const results = [{
    filePath: '/my-project/test.js',
    messages: [
    {
        ruleId: 'no-undef',
        severity: 2,
        message: "'c' is not defined.",
        line: 3,
        column: 9,
        nodeType: 'Identifier',
        messageId: 'undef',
        endLine: 3,
        endColumn: 10
    }],
    suppressedMessages: [],
    errorCount: 1,
    fatalErrorCount: 0,
    warningCount: 0,
    fixableErrorCount: 0,
    fixableWarningCount: 0,
    source: '/* eslint no-undef: "error" */\n\na = b + c;\n',
    usedDeprecatedRules: []
}];

// unmatched
{}
```

Reply:
If I understand correctly, this means removing the first `count` messages with a particular `rule-id` from each result, where `count` is the suppression count. So, in the above example `a` and `b` would be suppressed but not `c`, because in the input result, `a` and `b` are sorted first in order of location.

I think this behavior would be very surprising to users. This means that some rule violations would be reported only for (the bottom) part of a file.

I think it would be better to report all messages unfiltered if the suppression count is less than the number of messages, because in that situation we can't make a meaningful selection. So for each file and `rule-id` we should always report either all messages, or none of them.